src/bootstrap/src/utils/execution_context.rs

```diff
@@ -105,38 +107,107 @@ impl ExecutionContext {
         cmd.stdout(stdout.stdio());
         cmd.stderr(stderr.stdio());
 
-        let output = cmd.output();
+        let child = cmd.spawn().unwrap();
+
+        DeferredCommand::new(Some(child), stdout, stderr, command, Arc::new(self.clone()))
+    }
+
+    /// Execute a command and return its output.
+    /// Note: Ideally, you should use one of the BootstrapCommand::run* functions to
+    /// execute commands. They internally call this method.
+    #[track_caller]
+    pub fn run(
+        &self,
+        command: &mut BootstrapCommand,
+        stdout: OutputMode,
+        stderr: OutputMode,
+    ) -> CommandOutput {
+        self.start(command, stdout, stderr).wait_for_output()
+    }
+
+    fn fail(&self, message: &str, output: CommandOutput) -> ! {
+        if self.is_verbose() {
+            println!("{message}");
+        } else {
+            let (stdout, stderr) = (output.stdout_if_present(), output.stderr_if_present());
+            // If the command captures output, the user would not see any indication that
+            // it has failed. In this case, print a more verbose error, since to provide more
+            // context.
+            if stdout.is_some() || stderr.is_some() {
+                if let Some(stdout) = output.stdout_if_present().take_if(|s| !s.trim().is_empty()) {
+                    println!("STDOUT:\n{stdout}\n");
+                }
+                if let Some(stderr) = output.stderr_if_present().take_if(|s| !s.trim().is_empty()) {
+                    println!("STDERR:\n{stderr}\n");
+                }
+                println!("Command has failed. Rerun with -v to see more details.");
+            } else {
+                println!("Command has failed. Rerun with -v to see more details.");
+            }
+        }
+        exit!(1);
+    }
+}
+
+pub struct DeferredCommand<'a> {
+    process: Option<Child>,
+    command: &'a mut BootstrapCommand,
+    stdout: OutputMode,
+    stderr: OutputMode,
+    exec_ctx: Arc<ExecutionContext>,
+}
+
+impl<'a> DeferredCommand<'a> {
+    pub fn new(
+        child: Option<Child>,
+        stdout: OutputMode,
+        stderr: OutputMode,
+        command: &'a mut BootstrapCommand,
+        exec_ctx: Arc<ExecutionContext>,
+    ) -> Self {
+        DeferredCommand { process: child, stdout, stderr, command, exec_ctx }
+    }
+
+    pub fn wait_for_output(mut self) -> CommandOutput {
+        if self.process.is_none() {
+            return CommandOutput::default();
+        }
+        let output = self.process.take().unwrap().wait_with_output();
+
+        let created_at = self.command.get_created_location();
+        let executed_at = std::panic::Location::caller();
```

Comment:
This line has to be in `start`, because it is marked with `#[track_caller]`, and we want to track the place where the command was started, not necessarily where it was finished. We will thus need to store the `Location` in the deferred command.